File: python/Problem27.py
# coding:utf-8
"""
Euler discovered the remarkable quadratic formula:

$n^2 + n + 41$

It turns out that the formula will produce 40 primes for the consecutive
integer values $0 \le n \le 39$. However, when $n = 40, 40^2 + 40 + 41 = 40(40
+ 1) + 41$ is divisible by 41, and certainly when $n = 41, 41^2 + 41 + 41$ is
clearly divisible by 41.

The incredible formula $n^2 - 79n + 1601$ was discovered, which produces 80
primes for the consecutive values $0 \le n \le 79$. The product of the
coefficients, −79 and 1601, is −126479.

Considering quadratics of the form:

    $n^2 + an + b$, where $|a| < 1000$ and $|b| \le 1000$

    where $|n|$ is the modulus/absolute value of $n$
    e.g. $|11| = 11$ and $|-4| = 4$

Find the product of the coefficients, $a$ and $b$, for the quadratic expression
that produces the maximum number of primes for consecutive values of $n$,
starting with $n = 0$.
"""

import itertools


def is_prime(x):
    if x == 2:
        return True
    if x < 2 or x % 2 == 0:
        return False
    limit = int(x ** 0.5) + 1
    for i in range(3, limit):
        if x % i == 0:
            return False
    return True


def sieve(limit):
    search = bytearray([1] * limit)
    search[0] = 0
    search[1] = 0
    sqrt_limit = int(limit ** 0.5)
    for i in range(2, sqrt_limit):
        if search[i]:
            for j in range(i << 1, limit, i):
                search[j] = 0
    return itertools.compress(range(limit), search)


def count_quadratic_primes(a, b):
    count = 0
    for i in range(b):
        if is_prime(i * i + a * i + b):
            count += 1
        else:
            break
    return count


if __name__ == '__main__':
    longest = (0, 0)
    primes = list(sieve(1000))
    for a, b in itertools.product(range(-999, 1000, 2), primes):
        current = count_quadratic_primes(a, b)
        if longest[0] < current:
            longest = (current, a * b)
    print(longest[1])

File: java/Problem27.java
/*
 * Euler discovered the remarkable quadratic formula:
 *
 * n² + n + 41
 *
 * It turns out that the formula will produce 40 primes for the consecutive values
 * n = 0 to 39. However, when n = 40, 40^2 + 40 + 41 = 40(40 + 1) + 41 is
 * divisible by 41, and certainly when n = 41, 41² + 41 + 41 is clearly divisible
 * by 41.
 *
 * The incredible formula  n² − 79n + 1601 was discovered, which produces 80
 * primes for the consecutive values n = 0 to 79. The product of the coefficients,
 * −79 and 1601, is −126479.
 *
 * Considering quadratics of the form:
 *
 *     n² + an + b, where |a| < 1000 and |b| < 1000
 *
 *     where |n| is the modulus/absolute value of n
 *     e.g. |11| = 11 and |−4| = 4
 *
 * Find the product of the coefficients, a and b, for the quadratic expression
 * that produces the maximum number of primes for consecutive values of n,
 * starting with n = 0.
 */
import java.math.BigInteger;
import java.util.PrimitiveIterator;
import java.util.Spliterator;
import java.util.Spliterators;

public class Problem27 {
    private static class Primes implements Iterable<Integer> {
        private final int limit;
        private boolean[] isComposites;

        public Primes(int limit) {
            this.limit = limit;
        }

        @Override
        public PrimitiveIterator.OfInt iterator() {
            sieveIfNeeded();
            return new PrimesIterator();
        }

        @Override
        public Spliterator.OfInt spliterator() {
            return Spliterators.spliterator(iterator(),
                    (long) Math.floor(limit / Math.log(limit)),
                    Spliterator.DISTINCT | Spliterator.IMMUTABLE | Spliterator.NONNULL | Spliterator.ORDERED | Spliterator.SIZED | Spliterator.SORTED);
        }

        private synchronized void sieveIfNeeded() {
            if (isComposites == null) {
                isComposites = new boolean[limit];
                isComposites[0] = true;
                isComposites[1] = true;
                int sqrtLimit = (int)Math.sqrt(limit);
                for (int i = 2; i <= sqrtLimit; i++) {
                    if (!isComposites[i]) {
                        for (int j = i * i; j < limit; j += i) {
                            isComposites[j] = true;
                        }
                    }
                }
            }
        }

        private class PrimesIterator implements PrimitiveIterator.OfInt {
            private static final int NOT_FOUND = Integer.MIN_VALUE;
            private int next = 2;

            @Override
            public boolean hasNext() {
                return next != NOT_FOUND;
            }

            @Override
            public int nextInt() {
                int current = next;
                for (int i = current + 1; i < isComposites.length; i++) {
                    if (!isComposites[i]) {
                        next = i;
                        return current;
                    }
                }
                next = NOT_FOUND;
                return current;
            }
        }
    }

    private static class QuadraticFormula {
        private final int a;
        private final int b;

        public QuadraticFormula(int a, int b) {
            this.a = a;
            this.b = b;
        }

        public int solve(int n) {
            return n * n + a * n + b;
        }
    }

    public static void main(String[] args) {
        Primes primes = new Primes(1000);
        int longest = 0;
        int answer = 0;
        for (int a = -999; a <= 999; a += 2) {
            for (int b : primes) {
                QuadraticFormula formula = new QuadraticFormula(a, b);
                int n = 0;
                while (BigInteger.valueOf(formula.solve(n)).isProbablePrime(10)) {
                    n++;
                }
                if (longest < n) {
                    longest = n;
                    answer = a * b;
                }
            }
        }
        System.out.println(answer);
    }
}
